crates/kas-widgets/src/text.rs

```diff
@@ -39,19 +39,68 @@ mod Text {
             Text {
                 core: Default::default(),
                 text: theme::Text::new(T::default(), TextClass::Label(true)),
-                text_fn: Box::new(|_, data| data.into()),
+                text_fn: Box::new(|_, data, text| {
+                    let new_text = data.into();
+                    let changed = new_text != *text;
+                    if changed {
+                        *text = new_text;
+                    }
+                    changed
+                }),
+            }
+        }
+    }
+
+    impl<A> Text<A, String> {
+        /// Construct with an `str` accessor
+        pub fn new_str(as_str: impl Fn(&A) -> &str + 'static) -> Self {
+            Text {
+                core: Default::default(),
+                text: theme::Text::new(String::new(), TextClass::Label(true)),
+                text_fn: Box::new(move |_, data, text| {
+                    let s = as_str(data);
+                    let changed = *text != *s;
+                    if changed {
+                        *text = s.into();
+                    }
+                    changed
+                }),
             }
         }
     }
 
     impl Self {
-        /// Construct with a data binding
-        #[inline]
-        pub fn new(text_fn: impl Fn(&ConfigCx, &A) -> T + 'static) -> Self {
+        /// Construct with a generator function
+        ///
+        /// `gen_text` is called on each widget update to generate text from
+        /// input data.
+        pub fn new_gen(gen_text: impl Fn(&ConfigCx, &A) -> T + 'static) -> Self {
```

Comment:
Breaking change: `Text::new` is renamed `Text::new_gen`; additional constructors added: `new_str` and `new_update`